chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,809 @@
-                    return Err(Error::BlockNotFound);
-                }
+        if let RemoveTipResult::HitPermanentStore(id) = result {
+            let block_info = self
+                .get_block_info(&id)
+                .expect("parent block in permanent store not found");
+            let chain_length = block_info.chain_length() + 1;
 
-                tx.prepare_cached("insert into Tags (name, hash) values(?, ?)")
-                    .map_err(|err| Error::BackendError(Box::new(err)))?
-                    .execute(&[
-                        Value::Text(tag_name.to_string()),
-                        Value::Blob(block_hash.serialize_as_vec().unwrap()),
-                    ])
+            if self.get_blocks_by_chain_length(chain_length)?.is_empty() {
+                tips.insert(block_info.id(), &[])?;
             }
         }
-        .map_err(|err| Error::BackendError(Box::new(err)))?;
-
-        let row_id = tx.last_insert_rowid();
-        index
-            .add_tag(tag_name.to_owned(), block_hash, row_id as isize)
-            .map_err(|err| Error::BackendError(Box::new(err)))?;
-
-        tx.commit()
-            .map_err(|err| Error::BackendError(Box::new(err)))?;
 
         Ok(())
     }
 
-    pub fn get_tag(&mut self, tag_name: &str) -> Result<Option<B::Id>, Error> {
-        match self
-            .inner
-            .prepare_cached("select hash from Tags where name = ?")
-            .map_err(|err| Error::BackendError(Box::new(err)))?
-            .query_row(&[&tag_name], |row| Ok(blob_to_hash(row.get(0)?)))
-        {
-            Ok(s) => Ok(Some(s)),
-            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
-            Err(err) => Err(Error::BackendError(Box::new(err))),
-        }
-    }
-
-    /// Check whether a block exists.
-    pub fn block_exists(&mut self, block_hash: &B::Id) -> Result<bool, Error> {
-        let index = self.index.read().unwrap();
-        Self::block_exists_internal(&self.inner, &index, block_hash)
-    }
+    /// Check if the block with the given id exists.
+    pub fn block_exists(&self, block_id: &[u8]) -> Result<bool, Error> {
+        let info = self.volatile.open_tree(tree::INFO)?;
 
-    fn block_exists_internal(
-        connection: &Connection,
-        index: &ChainStorageIndex<B>,
-        block_hash: &B::Id,
-    ) -> Result<bool, Error> {
-        match Self::get_block_info_internal(connection, index, block_hash) {
-            Ok(_) => Ok(true),
-            Err(Error::BlockNotFound) => Ok(false),
-            Err(err) => Err(err),
+        if self.permanent.contains_key(block_id)? {
+            return Ok(true);
         }
+
+        info.get(block_id)
+            .map(|maybe_block| maybe_block.is_some())
+            .map_err(Into::into)
     }
 
-    // Determine whether block 'ancestor' is an ancestor of block 'descendent'
+    /// Determine whether block 'ancestor' is an ancestor of block 'descendent'
```

Comment:
As this is public API, please fix spelling to 'descendant'.

BTW, the doc does not refer to correct argument names and the convention is to have them in backticks.